### PR TITLE
add dry-run and verbose options

### PR DESCRIPTION
**Pull request by drostron**
a few samples:

    ➤ mackup backup --verbose --dry-run

    ➤ mackup restore --verbose --dry-run

    ➤ mackup uninstall --verbose --dry-run

### CODE REVIEW

**Comment by drostron** [comment]
rebased onto lra/mackup@b2e75a593b83ceaf3a7b8dd104142ac7c8ec6a28

**Comment by lra** [comment]
We switched to docopt since your PR, mind rebasing?

File: mackup/application.py [hunk 1]
```diff
@@ -45,18 +61,21 @@ def backup(self):
...
-            if ((os.path.isfile(filepath) or os.path.isdir(filepath))
-                and not (os.path.islink(filepath)
+            if ((os.path.isfile(home_filepath) or os.path.isdir(home_filepath))
+                and not (os.path.islink(home_filepath)
                          and (os.path.isfile(mackup_filepath)
                               or os.path.isdir(mackup_filepath))
-                         and os.path.samefile(filepath, mackup_filepath))):
+                         and os.path.samefile(home_filepath, mackup_filepath))):
+
+                print("Backing up {} ...".format("" if self.verbose else filename))
```
Comment (lra) [inline mackup/application.py]: Can you make an explicit if statement, I like to keep those construction to really dumb statements.

File: mackup/application.py [hunk 2]
```diff
@@ -80,18 +99,23 @@ def backup(self):
...
                     # Delete the file in the home
-                    utils.delete(filepath)
+                    utils.delete(home_filepath)
                     # Link the backuped file to its original place
-                    utils.link(mackup_filepath, filepath)
+                    utils.link(mackup_filepath, home_filepath)
+            elif self.verbose:
+              print("Doing nothing\n  {}\n  {}".format(
+                  home_filepath, 
+                  "is already backed up to\n  " + mackup_filepath
```
Comment (lra) [inline mackup/application.py]: "is already backed up to does not exist"? Feel free to put a more complex conditions here to get a wording that's gonna be easier to understand.

File: mackup/application.py [hunk 3]
```diff
@@ -124,7 +146,12 @@ def restore(self):
                                               home_filepath))
                 and utils.can_file_be_synced_on_current_platform(filename)):
 
-                print("Restoring {} ...".format(home_filepath))
+                print("Restoring {} ...".format("" if self.verbose else filename))
```
Comment (lra) [inline mackup/application.py]: again here, a simple "if verbose: ... else: ..." would make things easier to read.

File: mackup/application.py [hunk 4]
```diff
@@ -147,6 +174,12 @@ def restore(self):
                         utils.link(mackup_filepath, home_filepath)
                 else:
                     utils.link(mackup_filepath, home_filepath)
+            elif self.verbose:
+                print("Doing nothing\n  {}\n  {}".format(
+                    mackup_filepath,
+                    "already linked by\n  " + home_filepath
+                        if os.path.exists(home_filepath) else "does not exist"))
```
Comment (lra) [inline mackup/application.py]: same as https://github.com/lra/mackup/pull/277/files#r36595595

File: mackup/application.py [hunk 5]
```diff
@@ -174,10 +205,18 @@ def uninstall(self):
 
                 # Check if there is a corresponding file in the home folder
                 if os.path.exists(home_filepath):
-                    print("Reverting {} ...".format(home_filepath))
+                    print("Reverting {} ...".format(mackup_filepath if self.verbose else filename))
```
Comment (lra) [inline mackup/application.py]: `if verbose: ...`

File: mackup/main.py [hunk 6]
```diff
@@ -38,35 +48,45 @@ def main():
...
+
     if args.mode == BACKUP_MODE:
         # Check the env where the command is being run
         mckp.check_for_usable_backup_env()
 
-        # Backup each application
-        for app_name in mckp.get_apps_to_backup():
-            app = ApplicationProfile(mckp, app_db.get_files(app_name))
-            app.backup()
+        def backupApplication(app_name):
```
Comment (lra) [inline mackup/main.py]: Instead if decaring a function inside a function that your are gonna pass to another function, it'd be easier if this logic was the responsibility of the ApplicationProfile class.

File: mackup/main.py [hunk 7]
```diff
@@ -38,35 +48,45 @@ def main():
...
+        foreachApp(mckp.get_apps_to_backup(), backupApplication)
 
     elif args.mode == RESTORE_MODE:
         # Check the env where the command is being run
         mckp.check_for_usable_restore_env()
 
-        for app_name in mckp.get_apps_to_backup():
-            app = ApplicationProfile(mckp, app_db.get_files(app_name))
-            app.restore()
+        def restoreApplication(app_name):
```
Comment (lra) [inline mackup/main.py]: same here, this code should be the responsibility of the ApplicationProfile class, modifying things on the fly makes it harder to maintain.

File: mackup/main.py [hunk 8]
```diff
@@ -38,35 +48,45 @@ def main():
...
+          ApplicationProfile(mckp, app_db.get_files(app_name), args.dry_run, args.verbose).restore()
+
+        foreachApp(mckp.get_apps_to_backup(), restoreApplication)
 
     elif args.mode == UNINSTALL_MODE:
         # Check the env where the command is being run
         mckp.check_for_usable_restore_env()
 
-        if utils.confirm("You are going to uninstall Mackup.\n"
+        if args.dry_run or \
```
Comment (lra) [inline mackup/main.py]: PEP8 does not like `\`, can u use parens?

File: mackup/main.py [hunk 9]
```diff
@@ -38,35 +48,45 @@ def main():
...
+           utils.confirm("You are going to uninstall Mackup.\n"
                          "Every configuration file, setting and dotfile"
                          " managed by Mackup will be unlinked and moved back"
                          " to their original place, in your home folder.\n"
                          "Are you sure ?"):
-            for app_name in mckp.get_apps_to_backup():
-                app = ApplicationProfile(mckp, app_db.get_files(app_name))
-                app.uninstall()
+
+            def uninstallApp(app_name):
```
Comment (lra) [inline mackup/main.py]: Same as https://github.com/lra/mackup/pull/277/files#r36595681

**Comment by drostron** [comment]
Happy to rebase and address comments. I should have some time later this week.

**Comment by drostron** [comment]
Rebased and addressed comments.

**Comment by lra** [comment]
thanks!

**Pull request merged**